File: src/clients/mix-server-client/src/app/main-content/file-explorer/models/file-explorer-folder-node.ts
import {FileExplorerNode} from "./file-explorer-node";
import {FileExplorerNodeType} from "../enums/file-explorer-node-type";
import {NodePath} from "./node-path";

export class FileExplorerFolderNode implements FileExplorerNode {
  constructor(public path: NodePath,
              public type: FileExplorerNodeType,
              public exists: boolean,
              public creationTimeUtc: Date,
              public belongsToRoot: boolean,
              public belongsToRootChild: boolean,
              public parent: FileExplorerFolderNode | undefined | null) {
    this.disabled = !exists;
  }

  public disabled: boolean = false;

  public mdIcon: string = 'folder';

  public isEqual(node: FileExplorerNode | null | undefined): boolean {
    if (!node) {
      return false;
    }

    if (!(node instanceof FileExplorerFolderNode)) {
      return false;
    }

    return this.path.isEqual(node.path);
  }

  public static get Default(): FileExplorerFolderNode {
    return new FileExplorerFolderNode(
      NodePath.Default,
      FileExplorerNodeType.Folder,
      false,
      new Date(),
      false,
      false,
      null);
  }

  public copy(): FileExplorerFolderNode {
    return new FileExplorerFolderNode(
      this.path.copy(),
      this.type,
      this.exists,
      this.creationTimeUtc,
      this.belongsToRoot,
      this.belongsToRootChild,
      this.parent ? this.parent.copy() : null
    );
  }
}

File: src/api/MixServer.Domain/FileExplorer/Models/FileExplorerFolderNode.cs
using MixServer.Domain.FileExplorer.Entities;
using MixServer.Domain.FileExplorer.Enums;

namespace MixServer.Domain.FileExplorer.Models;

public class FileExplorerFolderNode(
    NodePath path,
    FileExplorerNodeType type,
    bool exists,
    DateTime creationTimeUtc,
    bool belongsToRoot,
    bool belongsToRootChild,
    IFileExplorerFolderNode? parent)
    : IFileExplorerFolderNode
{
    public NodePath Path { get; } = path;
    public FileExplorerNodeType Type { get; } = type;
    public bool Exists { get; } = exists;
    public DateTime CreationTimeUtc { get; } = creationTimeUtc;
    public bool BelongsToRoot { get; } = belongsToRoot;
    public bool BelongsToRootChild { get; } = belongsToRootChild;
    public IFileExplorerFolderNode? Parent { get; } = parent;
}

public class FileExplorerFolderNodeWithEntity(
    IFileExplorerFolderNode node,
    FileExplorerFolderNodeEntity entity)
    : FileExplorerFolderNode(node.Path, node.Type, node.Exists, node.CreationTimeUtc, node.BelongsToRoot, node.BelongsToRootChild, node.Parent),
      IFileExplorerFolderNodeWithEntity
{
    public FileExplorerFolderNodeEntity Entity { get; } = entity;
}